docs/DeferredSignals.md

```diff
@@ -0,0 +1,271 @@
+"uninterruptible" code section, which if the section is fairly small then it might cost more to ensure state integrity than doing the work at all.
+
+### Cooperative signal deferring
+A new solution is to defer asynchronous signals if they are caught inside an uninterruptible section. At the basic level, we increment a reference
+counter going in to the "uninterruptible" section, and then decrement the reference counter once we leave. This way when the signal handler receives a
+signal, it can check that thread's reference counter, store the `siginfo_t` to an array/stack object, and return to the same code segment to be handled
+later.
+
+The trick to how this works is the cost it takes to check if a signal occured, and then handling the signal safely. Ideally no signal has happened in
+the "uninterruptible" code section, so the check needs to be as cheap as possible to ensure no cost in the case that no signal has occured.
+
+The trick is that FEX maintains two memory regions for tracking deferred signals **per thread**.
+
+#### 1st memory region
+
+This region is FEX's InternalThreadState object, which is always resident for each guest thread and usually inside a register inside the JIT.
+Inside this object is where the reference counter for "uninterruptible" code segments lives. It is specifically a reference counter since these
+code segments may nest inside each other and we can only interrupt with a signal if the counter is zero.
+
+This reference counter is thread local and won't be read by any other threads, so it can be a non-atomic increment and decrement.
+Meaning it is usually three instructions (on ARM64) to increment and decrement.
+
+```cpp
+MoveableNonatomicRefCounter<uint64_t> DeferredReferenceCounter;
+MoveableNonatomicRefCounter<uint64_t> *DeferredSignalHandlerPagePtr;
+```
+
+#### 2nd memory region
+
+This memory region is a single page of memory that is allocated per thread. A pointer to this region exists inside the InternalThreadState object just
+after the reference counter. This allows the JIT code to quickly load the pointer to this region after modifying the reference counter.
+
+This is the tricky memory page where we determine if we have any deferred signals to process. I say it's tricky because after FEX leaves an
+"uninterruptible" code section, it will decrement the reference counter, and then try to store in to the first byte of this page. In the case that no
+signal has been deferred, this memory store will progress without issue, consuming only two instructions in the process.
+
+In the case that a signal **HAS** been deferred, then the permissions on this page will be set to `PROT_NONE` and this memory store will cause a
+SIGSEGV. Then FEX's signal handler will check to see if the access was this special page and start the deferred signal mechanisms. This means that a
+deferred signal check is only five instructions in total.
```

Comment:
The mechanism of writing to this page is already outlined below, so this section can be more succinctly summarized like this:

```suggestion
This memory region is a single page of memory that is allocated per thread. Its purpose is to trigger a SIGSEGV when FEX leaves an "uninterruptible" section if a signal has been deferred. FEX's signal handler will check if the faulting address is in this special page and subsequently starts the deferred signal mechanisms.
```

Reply:
Resolved.